prov/util/src/util_mr_cache.c

```diff
@@ -0,0 +1,1139 @@
+	struct ofi_mr_cache_entry *entry;
+	struct ofi_subscription *subscription;
+	RbtIterator iter;
+
+	if (!UTIL_MR_CACHE_IS_LAZY_DEREG(cache))
+		return;
+
+	while ((subscription = ofi_monitor_get_event(&cache->mem_nq))) {
+		entry = container_of(subscription, struct ofi_mr_cache_entry,
+				     subscription);
+		if (ofi_atomic_get32(&entry->ref_cnt) != 0) {
+			/* First, warn that this might be a problem.*/
+			if (!util_mr_cache_notifier_warned &&
+			    !util_mr_cache_entry_is_merged(entry)) {
+				FI_WARN(UTIL_MR_CACHE_GET_PROV(cache), FI_LOG_MR,
+					"Registered memory region includes unmapped pages."
+					" Have you freed memory w/o closing the memory region?\n");
+				util_mr_cache_notifier_warned = 1;
+			}
+
+			FI_DBG(UTIL_MR_CACHE_GET_PROV(cache), FI_LOG_MR,
+			       "Marking unmapped entry (%p) as retired %"PRIu64":%"PRIu64"\n",
+			       entry, entry->region.address, entry->region.length);
+
+			util_mr_cache_entry_set_unmapped(entry);
+
+			if (util_mr_cache_entry_is_retired(entry))
+				/* Nothing to do */
+				break;
+
+			/* Retire this entry (remove from inuse tree) */
+			util_mr_cache_entry_set_retired(entry);
+			iter = rbtFind(cache->inuse_tree, &entry->region);
+			if (OFI_LIKELY(iter != NULL)) {
+				ret = rbtErase(cache->inuse_tree, iter);
+				if (OFI_UNLIKELY(ret != RBT_STATUS_OK)) {
+					FI_WARN(UTIL_MR_CACHE_GET_PROV(cache), FI_LOG_MR,
+						"Unmapped entry could not be removed from in usetree.\n");
+					abort();
+				}
+			} else {
+				/*  The only way we should get here is if we're in the
+				 *  middle of retiring this entry.  Not sure if this
+				 *  is worth a separate warning from the one above.
+				 */
+			}
+
+			break;
+		} else {
+			util_mr_cache_entry_set_unmapped(entry);
+			iter = rbtFind(cache->stale_tree, &entry->region);
+			if (!iter)
+				break;
+
+			ret = rbtErase(cache->stale_tree, iter);
+			if (OFI_UNLIKELY(ret != RBT_STATUS_OK)) {
+				FI_WARN(UTIL_MR_CACHE_GET_PROV(cache), FI_LOG_MR,
+					"Unmapped entry could not be removed from stale tree.\n");
+				abort();
+			}
+
+			FI_DBG(UTIL_MR_CACHE_GET_PROV(cache), FI_LOG_MR,
+			       "Removed unmapped entry (%p) from stale tree %"PRIu64":%"PRIu64"\n",
+			       entry, entry->region.address, entry->region.length);
+
+			util_mr_cache_lru_remove(cache, entry);
+			FI_DBG(UTIL_MR_CACHE_GET_PROV(cache), FI_LOG_MR,
+			       "Removed unmapped entry (%p) from lru list %"PRIu64":%"PRIu64"\n",
+			       entry, entry->region.address, entry->region.length);
+			cache->stale_elem--;
+
+			util_mr_cache_entry_destroy(cache, entry);
+
+			break;
+		}
+	}
+	if (OFI_UNLIKELY(ret != -FI_EAGAIN))
+		/* Should we do something else here? */
+		FI_WARN(UTIL_MR_CACHE_GET_PROV(cache), FI_LOG_MR,
+			"`get event` returned error: %s\n",
+			fi_strerror(-ret));
+
+	return;
+}
+
+static int
+util_mr_cache_notifier_monitor(struct ofi_mr_cache *cache,
+			       struct ofi_mr_cache_entry *entry)
+{
+
+    if (!UTIL_MR_CACHE_IS_LAZY_DEREG(cache))
```

Comment:
indentation seems inconsistent in this function

Reply:
Thanks. Done